### ISSUE TITLE
Consider API break for `hal::can::settings`

### ISSUE THREAD

**Issue by kammce** [libhal/libhal#40]
# EDIT: We gonna break the API

Because all other options are just so much more confusing.

---

Looking at the settings object, it feels strange that the propagation, phase segments and sjw are needed to set the baud rate. To get this value right you'd need to know the device's peripheral clock which feels like a layer too deep for the generic interface. It makes sense to potentially break the API of `hal::can::settings` or change the meaning of fields.

## The breaking path

The break would be to delete these fields and increment the major version of libhal (hopefully with other critical changes)

## The deprecation path

We can deprecate these fields and have most code ignore them. The only issue is, what do users that want to tune these values do to make that happen? There are rare cases where a user may really want to change these values and in this case, wouldn't this struct be misleading?

## The optional path

We cannot make these optional as that would break the ABI and potentially the API. But we could create an "optional" value. We can consider a value of 0 to mean "not specified" for every field except obviously baud rate. That way, if code encounters a zero for any of the tuning parameters it just ignores all of them and calculates something that works. 

## The extended API path

We add another API called `can::driver_baud_rate(hertz p_baud_rate)` which configures the baud rate of the can bus to the selected baud rate or throws an exception. We keep `configure(settings&)` for backwards compatibility and add the new API. To not cause an API break, we will need to make sure this isn't a pure virtual function. 

## The utility path

Create a until in `libhal-util/can.hpp` that you pass a can bus into along with a baud rate and it performs the calculation for the settings you need and calls `.configure()` on the can device if it was successful. Throws an exception otherwise.

I think I prefer this option because at the very least, we don't break anyone AND users keep that control if it actually turns out to be useful.

**Comment by kammce** [comment]
```C++
/**
   * @brief Generic settings for a can peripheral
   *
   * CAN Bit Quanta Timing Diagram of:
   *
   *                               | <--- sjw ---> |
   *         ____    ______    __________    __________
   *      _/ SYNC \/  PROP  \/   PHASE1   \/   PHASE2   \_
   *       \______/\________/\____________/\____________/
   *                                       ^ Sample point
   */
  struct settings
  {
    /**
     * @brief Bus clock rate in hertz
     *
     */
    hertz baud_rate = 100.0_kHz;

    /**
     * @brief Sync Segment (always 1qt)
     *
     * Initial sync transition, the start of a CAN bit
     */
    static constexpr std::uint8_t sync_segment = 1;

    /**
     * @brief Propagation Delay (1qt ... 8qt)
     *
     * Propagation time It is used to compensate for signal delays across the
     * network.
     */
    std::uint8_t propagation_delay = 3;

    /**
     * @brief Length of Phase Segment 1 (1qt ... 8qt)
     *
     * Determines the bit rate, phase segment 1 acts as a buffer that can be
     * lengthened to resynchronize with the bit stream via the
     * synchronization_jump_width. Includes propagation delay
     */
    std::uint8_t phase_segment1 = 3;

    /**
     * @brief Length of Phase Segment 2 (1qt ... 8qt)
     *
     * Determines the bit rate and is like phase segment 1 and occurs after the
     * sampling point. Phase segment 2 can be shortened to resynchronize with
     * the bit stream via the synchronization_jump_width.
     */
    std::uint8_t phase_segment2 = 3;

    /**
     * @brief Synchronization jump width (1qt ... 4qt)
     *
     * This is the maximum time by which the bit sampling period may be
     * lengthened or shortened during each cycle to adjust for oscillator
     * mismatch between nodes.
     *
     * This value must be smaller than phase_segment1 and phase_segment2
     */
    std::uint8_t synchronization_jump_width = 1;
  };
```

The new struct would be this if we broke the API:

```C++
  struct settings
  {
    /// @brief Bus clock rate in hertz
    hertz baud_rate = 100.0_kHz;
  };
```

By keeping the settings as they are though, we make it harder to use configure driver because the user will have to hand write or use other util APIs to generate the can settings then set whatever new items we add to the settings.

**Comment by kammce** [comment]
The one issue with the utility path is that you'd still need to know the operating frequency of the can peripheral in order to actually configure its baud rate. 

This also makes the can object's `configure()` function defacto unusable by anything generic because it would need this info. A gross work around could be to add

```C++
virtual hertz input_frequency() = 0;
```

Now the utility would work as needed, but would still be an API break.

**Comment by kammce** [comment]
After some thought, I think any cover up of this would simply cause a lot of confusion. I feel confident that making this breaking change will break so few bits of code that libraries could technically allow for this `libhal/[^3.0.0 || ^4.0.0]`. But a clean break and migration of all the projects should be short work since most code doesn't touch `hal::can`.

Affected packages would be:

1. libhal
2. libhal-util
3. libhal-mock
4. libhal-canrouter
5. libhal-lpc40
6. libhal-stm32f1

Most projects do not use the tuning parameters so they'll probably not even break.

**Comment by kammce** [comment]
Resolved by #41
